internal_docs/generate_release_changelog/main.go

```diff
@@ -0,0 +1,255 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/cheggaaa/pb/v3"
+	"github.com/google/go-github/v28/github"
+	"github.com/spf13/pflag"
+	"golang.org/x/oauth2"
+)
+
+const uncategorizedSectionTitle = "Uncategorized"
+
+func getCommitsInRange(ctx context.Context, ghClient *github.Client, gitHubOrg string, gitHubRepo string,
+	fromRevision string, toRevision string) ([]github.RepositoryCommit, error) {
+
+	// Make sure that the "from" revision is behind "to" and obtain the commits
+	comparison, _, err := ghClient.Repositories.CompareCommits(ctx, gitHubOrg, gitHubRepo, fromRevision, toRevision)
+	if err != nil {
+		return nil, fmt.Errorf("cannot compare commits from %q and to %q: %s", fromRevision, toRevision, err)
+	}
+	if comparison.GetStatus() != "ahead" || comparison.GetBehindBy() != 0 {
+		return nil, fmt.Errorf("'from' revision (%s) is not behind 'to' revision (%s)", fromRevision, toRevision)
+	}
+	return comparison.Commits, nil
+}
+
+func findPullRequestsInCommits(ctx context.Context, ghClient *github.Client, gitHubOrg string, gitHubRepo string,
+	commits []github.RepositoryCommit, commitProcessHook func()) ([]*github.PullRequest, error) {
+	// Store pull requests in visiting order, preventing duplicates
+	prs := []*github.PullRequest{}
+	visitedPRs := map[int]struct{}{}
+	for _, commit := range commits {
+		commitProcessHook()
+		pullRequestListOptions := &github.PullRequestListOptions{}
+		// Note: Unfortunately this will return all the PRs whose head branch contains the commit
+		// (i.e. not just the PR which incorporates the commit to the base repo).
+		// We deal with this by discarding duplicates.
+		prsForCommit, _, err := ghClient.PullRequests.ListPullRequestsWithCommit(ctx,
```

Comment:
Calling this on every commit is very slow (that's why I added the progress bar). As an alternative we could parse the commit messages looking for `Merge` messages (and messages ending with `#number` for rebased PRs, which don't have a Merge message). However, I deemed that to be a bit too fragile.

Reply:
You can look for commits with more than one parent. That will catch all PR merges except those that are fast-forwards, which should be none.

Reply:
> That will catch all PR merges except those that are fast-forwards, which should be none.

We explicitly disable that?

Reply:
Yes I believe so. (But of course, that may change, you never know)

Reply:
Yes, I would rather not make assumptions.